### PR TITLE
Functional tests - Add new test "Customers > Outstanding page" Pagination - Sort - Filter

### PR DESCRIPTION
**Pull request by khouloudbelguith**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | Add a new test for the "Customers > Outstanding page" Pagination - Sort - Filter
| Type?             | refacto
| Category?         | TE
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | no
| How to test?      | HEADLESS=false TEST_PATH="functional/BO/04*/03*" URL_FO="http://shop.com/" npm run test:specific
| Possible impacts? | no

### CODE REVIEW

File: tests/UI/commonTests/BO/customers/createDeleteCustomer.js [hunk 1]
```diff
@@ -69,6 +69,55 @@ function createCustomerTest(customerData, baseContext = 'commonTests-createCusto
...
+
+    after(async () => {
+      await helper.closeBrowserContext(browserContext);
+    });
+
+    it('should login in BO', async function () {
+      await loginCommon.loginBO(this, page);
+    });
+
+    it('should go \'Customers > Customers\' page', async function () {
```
Comment (nesrineabdmouleh) [inline tests/UI/commonTests/BO/customers/createDeleteCustomer.js]:
```suggestion
    it('should go to \'Customers > Customers\' page', async function () {
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  });
+  describe('Pagination next and previous', async () => {
+    it('should go to BO > Customers > Outstanding page', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'goToOutstandingPage', baseContext);
+
+      await dashboardPage.goToSubMenu(
+        page,
+        dashboardPage.customersParentLink,
+        dashboardPage.outstandingLink,
+      );
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js:120]:
```suggestion
      );
      
      pageTitle = await outstandingPage.getPageTitle(page);
      await expect(pageTitle).to.contains(outstandingPage.pageTitle);
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+// Import faker data
+const CustomerFaker = require('@data/faker/customer');
+const AddressFaker = require('@data/faker/address');
+
+// Import demo data
+const {PaymentMethods} = require('@data/demo/paymentMethods');
+const {Statuses} = require('@data/demo/orderStatuses');
+
+const baseContext = 'functional_BO_customers_outstanding_filter_and_sort';
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js]:
```suggestion
const baseContext = 'functional_BO_customers_outstanding_sortFilterOutstanding;
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        } else {
+          await expect(sortedTable).to.deep.equal(expectedResult.reverse());
+        }
+      });
+    });
+  });
+
+  // Post-Condition: Delete created customers by bulk action
+  customersData.forEach((customerData, index = 1) => {
+    bulkDeleteCustomersTest('email', customerData.email, `${baseContext}_postTest_${index}`);
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js]:
```suggestion
  customersData.forEach((customerData, index ) => {
    bulkDeleteCustomersTest('email', customerData.email, `${baseContext}_postTest_${index + 1}`);
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+const baseContext = 'functional_BO_customers_outstanding_pagination';
+
+let browserContext;
+let page;
+
+// Variable used to get the number of outstanding
+let numberOutstanding;
+
+// Const used to get the least number of outstanding to display pagination
+const firstPagination = 11;
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js]:
```suggestion
const numberOfOrdersToCreate = 11;
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  // before and after functions
+  before(async function () {
+    browserContext = await helper.createBrowserContext(this.browser);
+    page = await helper.newTab(browserContext);
+  });
+
+  after(async () => {
+    await helper.closeBrowserContext(browserContext);
+  });
+  describe('PRE-TEST: Create outstanding', async () => {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js]:
```suggestion
  describe('PRE-TEST: Create 11 orders on FO and change their status to payment accepted on BO ', async () => {
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    it('should login to BO', async function () {
+      await loginCommon.loginBO(this, page);
+    });
+
+    const creationTests = new Array(firstPagination).fill(0, 0, firstPagination);
+    creationTests.forEach((value, index) => {
+      createOrderByCustomerTest(orderByCustomerData, `${baseContext}_preTest_${index}`);
+
+      // Pre-condition: Update order status to payment accepted
+      describe('PRE-TEST: Update order status to payment accepted', async () => {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js]:
```suggestion
      describe('Update order status to payment accepted', async () => {
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+      );
+    });
+
+    it('should reset all filters and get the number of outstanding', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'resetFilterAndGetNumberOfOutstanding');
+
+      await outstandingPage.resetFilter(page);
+
+      numberOutstanding = await outstandingPage.getNumberOutstanding(page);
+      await expect(numberOutstanding).to.be.above(0);
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/04_pagination.js]:
```suggestion
      await expect(numberOutstanding).to.be.above(numberOfOrdersToCreate);
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+- Filter outstanding by: ID, Date, Customer, Company, Risk, Outstanding allowance
+- Sort outstanding by: ID, Date, Customer, Company, Outstanding allowance DESC and ASC
+Post-condition:
+- Delete created customers by bulk actions
+- Disable B2B
+*/
+describe('BO - Customers - Outstanding : Filter and sort the Outstanding table', async () => {
+  // Pre-Condition : Enable B2B
+  enableB2BTest(baseContext);
+
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js:69]:
```suggestion
before(async function () {
    browserContext = await helper.createBrowserContext(this.browser);
    page = await helper.newTab(browserContext);
  });

  after(async () => {
    await helper.closeBrowserContext(browserContext);
  });
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+  // before and after functions
+  before(async function () {
+    browserContext = await helper.createBrowserContext(this.browser);
+    page = await helper.newTab(browserContext);
+  });
+
+  after(async () => {
+    await helper.closeBrowserContext(browserContext);
+  });
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js]:
```suggestion
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+
+  // Go to the outstanding page
+  describe('Go to the outstanding page', async () => {
+    it('should go to BO > Customers > Outstanding page', async function () {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js]:
```suggestion
    it('should go to 'Customers > Outstanding' page', async function () {
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+      await testContext.addContextItem(this, 'testIdentifier', 'goToOutstandingPage', baseContext);
+
+      await dashboardPage.goToSubMenu(
+        page,
+        dashboardPage.customersParentLink,
+        dashboardPage.outstandingLink,
+      );
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js:147]:
```suggestion
      );
      
      const pageTitle = await outstandingPage.getPageTitle(page);
      await expect(pageTitle).to.contains(outstandingPage.pageTitle);
      
```

File: tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+    });
+    it('should reset filter and get the last outstanding ID', async function () {
```
Comment (nesrineabdmouleh) [inline tests/UI/campaigns/functional/BO/04_customers/03_outstanding/05_sortFilterOutstanding.js]:
```suggestion
    it('should reset filter and get the outstanding number', async function () {
```

**Review by nesrineabdmouleh — COMMENTED** [review]
Thank you @khouloudbelguith for your PR :)
Some suggestions for you !

**Comment by khouloudbelguith** [comment]
Hello @nesrineabdmouleh,

Thanks for the review.
Fix done.

Thanks!

**Review by nesrineabdmouleh — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]

**Pull request merged**